### DISCUSSION TITLE
Typesetting ragged-right?

### DISCUSSION THREAD

**Discussion by jslabovitz** [sile-typesetter/sile#1719] [Q&A] [answered]
I'm typesetting a long book (a memoir), and I'd like the lines of the main text to be set ragged-right. I still want paragraph indents and hyphenation, as well as occasional centered lines (such as headings or figures) or right-justified lines (like author attributions/signatures).

I've tried various combinations of using the `\raggedright` environment, but can't seem to find something that retains paragraph spacing and hyphenation settings.

Probably complicating the issue is that this is a complex document, with a custom (but not complicated) `book` subclass, an XML 'master' document, and most content in Markdown.

Is there an accepted way to do this?

**Comment by Omikhleia** [comment]
The default "raggedright" environment (= "ragged" environment called with option "ragged=right") kills the paragraph indent https://github.com/sile-typesetter/sile/blob/5344cb5b6b8434fa12d55ebca9963842b486f3a6/classes/plain.lua#L316

But what prevents you from implementing your own environment that would keep it, and tune other things at convenience, if need be?

Simple proof of concept:

```
\begin[papersize=a7]{document}
\use[module=packages.lorem]
\font[size=13pt]
\set[parameter=linebreak.emergencyStretch, value=25%lw]% prefer underfilled lines over overfilled lines
\language[main=en]% Some language with hyphenation patterns

\lua{
  SILE.registerCommand("raggedright2", function (_, content)
    SILE.settings:temporarily(function ()
      SILE.settings:set("document.rskip", SILE.nodefactory.hfillglue())
      SILE.settings:set("typesetter.parfillskip", SILE.nodefactory.glue())
      SILE.settings:set("document.spaceskip", SILE.length("1spc", 0, 0))
      SILE.process(content)
      SILE.call("par")
    end)
  end)
}
\begin{raggedright2}
\lorem[words=20]. ThisIsAlongWordJustToSeeIfItBreaksSomewhereAsWeHaveHyphenations.

\lorem[words=20]
\end{raggedright2}
\end{document}
```

![image](https://user-images.githubusercontent.com/18075640/220987323-fb4c2575-327b-4238-a0c8-d531b43fba7f.png)

Obviously, a full hfillglue is so strong it will push the text maximally, so you will seldom get hyphenated words (other than when they are hardly avoidable). One could consider other approaches, but it enters into the deep field of fine tuning several other settings .... so I won't dabble into that uncertain direction yet ;-) -- but just note the SILE manual explains all the relevant parameters (look, for instance, for `\document.rskip`, `shaper.variablespaces`, `typesetter.parfillskip`, and the various `linebreak.xxx` parameters).

**Comment by alerque** [comment, marked as answer]
If you have a book subclass the instantiator for that is probably the best place to set this up. Rather than wrapping your entire document in a `\raggedright` environment I would just copy the relevant settings from that command and put them in your document init and make them the defaults for the whole document:

```lua
class:_init()
	book._init(self)
	SILE.settings:set("document.rskip", SILE.nodefactory.hfillglue(), true)
end
```

I would also take that chance to set other values. You *might* want to set the spaceskip to not be stretchy at all:

```
	SILE.settings:set("document.spaceskip", SILE.length("1spc", 0, 0), true)
```

Then again you might actually want a little bit of stretch. I've done some projects using hybrid justification that looks ragged right but actually has a little bit of stretch to the spaces to and hyphenation so the edge is much less ragged than if no line stretching was allowed. To do that you would change the rskip to something with a moderate chunk of stretchiness, say a `SILE.nodefactory.glue("0 8em")`, then give your space skips some very moderate stretch factor as well. Depends on your taste and use case of course.

**Reply by jslabovitz** [reply]
I'm having a hard time coming up with values that result in the kind of 'hybrid justification' you're talking about. I don't suppose you have some actual :set() calls you could throw over the fence?

**Reply by alerque** [reply]
Sure, here's some to lob over the fence from a project I was just using it on:

```
SILE.settings:set("document.rskip", SILE.nodefactory.glue("0 plus 6spc"))
SILE.settings:set("shaper.spaceenlargementfactor", 5/4, true)
SILE.settings:set("shaper.spacestretchfactor", 1, true)
SILE.settings:set("shaper.spaceshrinkfactor", 1/3, true)
SILE.settings:set("linebreak.hyphenPenalty", 150)
SILE.settings:set("linebreak.finalHyphenDemerits", 600)
SILE.settings:set("linebreak.emergencyStretch", SILE.length("3spc"))
```

Untested in isolation on many other settings, but but copy and pasted out a project using the `cabook` class from CaSILE.

The logic here is to allow spaces to stretch and shrink *a little* but not as much as one might otherwise do to achieve full justification. Then instead of not having any rskip (as in full justification) or having infinitely stretchable rskip (as in ragged right) I give it up to about 6 spaces worth of rskip. That allows it to be a little ragged but also stretch lines a little. Then I adjusted the hyphenation penalties so hyphenation can still happen when that 6 spaces plus line stretch isn't enough, but isn't as likely to happen as in a normal full justification. Finally when that still isn't enough it gets up to 3 more spaces worth to space out words past the normal space enlargement factor.

Note the optimal values here are very sensitive to the overall width of the line & base font size. Here's an example of a few paragraphs typeset with those settings:

![image](https://user-images.githubusercontent.com/173595/221028535-d8419d35-2d0d-41f0-aaad-2fed03e48c4c.png)

Note I had to flip through 30 pages or so to find an example of a lite that actually got hyphenated.

**Reply by alerque** [reply]
P.S. You'll probably want to add a third argument `true` to each of those sets to make them the default. I'm using nested wrappers with various classes applied so making the settings default isn't appropriate, but for the use case I suggested to you it probably would be.

**Comment by jslabovitz** [comment]
Perfect! This is exactly what I was looking for. I just didn't understand what it was called, and where exactly to set it.

**Comment by jslabovitz** [comment]
Cheers! Thank you.
